File: app/src/main/java/me/ykrank/s1next/data/pref/DownloadPreferencesManager.java
package me.ykrank.s1next.data.pref;

import android.support.annotation.IntDef;

import com.bumptech.glide.load.Key;
import com.bumptech.glide.signature.ObjectKey;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import me.ykrank.s1next.data.Wifi;
import me.ykrank.s1next.util.DateUtil;

/**
 * A manager manage the download preferences that are associated with settings.
 */
public final class DownloadPreferencesManager {

    private final DownloadPreferencesRepository mDownloadPreferencesProvider;
    private final Wifi mWifi;

    public DownloadPreferencesManager(DownloadPreferencesRepository downloadPreferencesProvider, Wifi wifi) {
        this.mDownloadPreferencesProvider = downloadPreferencesProvider;
        this.mWifi = wifi;
    }

    public int getTotalImageCacheSize() {
        return TotalDownloadCacheSize.getByte(Integer.parseInt(
                mDownloadPreferencesProvider.getTotalImageCacheSizeString()));
    }

    public int getTotalDataCacheSize() {
        return TotalDownloadCacheSize.getMByte(Integer.parseInt(
                mDownloadPreferencesProvider.getTotalDataCacheSizeString()));
    }

    public boolean isAvatarsDownload() {
        int avatarDownloadStrategy = Integer.parseInt(mDownloadPreferencesProvider.getAvatarsDownloadStrategyString());
        return DownloadStrategyInternal.isDownload(avatarDownloadStrategy, mWifi.isWifiEnabled());
    }

    public boolean isHighResolutionAvatarsDownload() {
        int avatarResolutionStrategy = Integer.parseInt(mDownloadPreferencesProvider.getAvatarResolutionStrategyString());
        return AvatarResolutionStrategy.isHigherResolutionDownload(avatarResolutionStrategy, mWifi.isWifiEnabled());
    }

    public Key getAvatarCacheInvalidationIntervalSignature() {
        int interval = Integer.parseInt(mDownloadPreferencesProvider.getAvatarCacheInvalidationIntervalString());
        return AvatarCacheInvalidationInterval.getSignature(interval);
    }

    /**
     * Checks whether we need to download images.
     */
    public boolean isImagesDownload() {
        int imageDownloadStrategy = Integer.parseInt(mDownloadPreferencesProvider.getImagesDownloadStrategyString());
        return DownloadStrategyInternal.isDownload(imageDownloadStrategy, mWifi.isWifiEnabled());
    }

    /**
     * Checks whether we need to monitor the Wi-Fi status.
     * We needn't monitor the Wi-Fi status if we needn't/should
     * download avatars or images.
     */
    public boolean needMonitorWifi() {
        int avatarDownloadStrategy = Integer.parseInt(mDownloadPreferencesProvider.getAvatarsDownloadStrategyString());
        int imageDownloadStrategy = Integer.parseInt(mDownloadPreferencesProvider.getImagesDownloadStrategyString());
        return avatarDownloadStrategy == DownloadStrategyInternal.WIFI
                || imageDownloadStrategy == DownloadStrategyInternal.WIFI;
    }

    private static class TotalDownloadCacheSize {
        private static final int LOW = 32;
        private static final int NORMAL = 64;
        private static final int HIGH = 128;

        private static final int[] SIZE = new int[]{LOW, NORMAL, HIGH};

        public static int getByte(int index) {
            return SIZE[index] * 1000 * 1000;
        }

        public static int getMByte(int index) {
            return index < 0 || index >= SIZE.length ? SIZE[0] : SIZE[index];
        }
    }

    private static class DownloadStrategyInternal {
        private static final int NOT = 0;
        private static final int WIFI = 1;
        private static final int ALWAYS = 2;

        @IntDef({NOT, WIFI, ALWAYS})
        @Retention(RetentionPolicy.SOURCE)
        @interface DownloadStrategy {
        }

        private static boolean isDownload(int downloadStrategy, boolean hasWifi) {
            return downloadStrategy == WIFI && hasWifi
                    || downloadStrategy == ALWAYS;
        }
    }

    private static class AvatarResolutionStrategy {
        private static final int LOW = 0;
        private static final int HIGH_WIFI = 1;
        private static final int HIGH = 2;

        private static boolean isHigherResolutionDownload(int avatarResolutionStrategy, boolean hasWifi) {
            return avatarResolutionStrategy == HIGH_WIFI && hasWifi
                    || avatarResolutionStrategy == HIGH;
        }
    }

    private static class AvatarCacheInvalidationInterval {
        private static final String EVERY_DAY = DateUtil.today();
        private static final String EVERY_WEEK = DateUtil.dayOfWeek();
        private static final String EVERY_MONTH = DateUtil.dayOfMonth();

        private static final String[] VALUES = new String[]{EVERY_DAY, EVERY_WEEK, EVERY_MONTH};

        /**
         * Gets a string signature in order to invalidate avatar every day/week/month.
         *
         * @return A {@link Key} representing the string signature
         * of date that will be mixed in to the cache key.
         */
        private static Key getSignature(int index) {
            return new ObjectKey(index < 0 || index >= VALUES.length ? VALUES[0] : VALUES[index]);
        }
    }
}

File: app/src/main/java/me/ykrank/s1next/data/pref/DownloadPreferencesManager.kt
package me.ykrank.s1next.data.pref

import androidx.annotation.IntDef
import com.bumptech.glide.load.Key
import com.bumptech.glide.signature.ObjectKey
import com.github.ykrank.androidtools.util.DateUtil
import me.ykrank.s1next.data.Wifi

/**
 * A manager manage the download preferences that are associated with settings.
 */
class DownloadPreferencesManager(
    val downloadPreferences: DownloadPreferences,
    private val mWifi: Wifi
) {

    val totalImageCacheSize: Long
        get() = TotalDownloadCacheSize.getByte(downloadPreferences.totalImageCacheSizeIndex)

    val totalDataCacheSize: Int
        get() =  downloadPreferences.totalDataCacheSize

    var downloadPath: String?
        get() = downloadPreferences.downloadPath
        set(path) {
            downloadPreferences.downloadPath = path
        }

    val netCacheEnable: Boolean
        get() = downloadPreferences.netCacheEnable

    val isAvatarsDownload: Boolean
        get() = DownloadStrategyInternal.isDownload(
            downloadPreferences.avatarsDownloadStrategyIndex,
            mWifi.isWifiEnabled
        )

    val avatarCacheInvalidationIntervalSignature: Key
        get() = AvatarCacheInvalidationInterval.getSignature(downloadPreferences.avatarCacheInvalidationInterval)

    /**
     * Checks whether we need to download images.
     */
    val isImagesDownload: Boolean
        get() = DownloadStrategyInternal.isDownload(
            downloadPreferences.imagesDownloadStrategyIndex,
            mWifi.isWifiEnabled
        )

    val postMaxImageShow: Int
        get() = downloadPreferences.postMaxImageShow

    /**
     * Checks whether we need to monitor the Wi-Fi status.
     * We needn't monitor the Wi-Fi status if we needn't/should
     * download avatars or images.
     */
    fun needMonitorWifi(): Boolean {
        val avatarDownloadStrategy = downloadPreferences.avatarsDownloadStrategyIndex
        val imageDownloadStrategy = downloadPreferences.imagesDownloadStrategyIndex
        return avatarDownloadStrategy == DownloadStrategyInternal.WIFI || imageDownloadStrategy == DownloadStrategyInternal.WIFI
    }

    private object TotalDownloadCacheSize {
        private val LOW = 128
        private val NORMAL = 256
        private val HIGH = 512

        private val SIZE = intArrayOf(LOW, NORMAL, HIGH)

        fun getByte(index: Int): Long {
            return SIZE[index] * 1024L * 1024
        }

        fun getMByte(index: Int): Int {
            return if (index < 0 || index >= SIZE.size) SIZE[0] else SIZE[index]
        }
    }

    private object DownloadStrategyInternal {
        const val NOT = 0
        const val WIFI = 1
        const val ALWAYS = 2

        @IntDef(NOT, WIFI, ALWAYS)
        @Retention(AnnotationRetention.SOURCE)
        internal annotation class DownloadStrategy

        fun isDownload(@DownloadStrategy downloadStrategy: Int, hasWifi: Boolean): Boolean {
            return downloadStrategy == WIFI && hasWifi || downloadStrategy == ALWAYS
        }
    }

    private object AvatarResolutionStrategy {
        const val LOW = 0
        const val HIGH_WIFI = 1
        const val HIGH = 2

        @IntDef(LOW, HIGH_WIFI, HIGH)
        @Retention(AnnotationRetention.SOURCE)
        internal annotation class AvatarStrategy

        fun isHigherResolutionDownload(
            @AvatarStrategy avatarResolutionStrategy: Int,
            hasWifi: Boolean
        ): Boolean {
            return avatarResolutionStrategy == HIGH_WIFI && hasWifi || avatarResolutionStrategy == HIGH
        }
    }

    private object AvatarCacheInvalidationInterval {
        private val EVERY_DAY = DateUtil.today()
        private val EVERY_WEEK = DateUtil.dayOfWeek()
        private val EVERY_MONTH = DateUtil.dayOfMonth()

        private val VALUES = arrayOf(EVERY_DAY, EVERY_WEEK, EVERY_MONTH)

        /**
         * Gets a string signature in order to invalidate avatar every day/week/month.

         * @return A [Key] representing the string signature
         * * of date that will be mixed in to the cache key.
         */
        fun getSignature(index: Int): Key {
            return ObjectKey(if (index < 0 || index >= VALUES.size) VALUES[0] else VALUES[index])
        }
    }
}
